Permissions.py

```diff
@@ -0,0 +1,197 @@
+                # variable found
+                variable = part[2:-1]
+                try:
+                    expansions = self.m_variables[variable]
+                except KeyError:
+                    raise Exception("Undeclared variable '{}' encountered in profile".format(variable))
+
+                new_ret = []
+
+                for p in ret:
+                    for value in expansions:
+                        new_ret.append(os.path.sep.join([p, value]))
+
+                ret = new_ret
+            elif not part:
+                # a leading slash, ignore
+                continue
+            else:
+                # a regular, fixed string
+                ret = [os.path.sep.join([p, part]) for p in ret]
+
+        if path.endswith(os.path.sep):
+            # restore trailing slashes since they signify that we
+            # expect a directory
+            ret = [p + os.path.sep for p in ret]
+
+        return ret
+
+
+class PermissionsParser:
+
+    def __init__(self, var_handler, profile_path):
+
+        self.m_var_handler = var_handler
+        self.m_entries = {}
+
+        with open(profile_path) as fd:
+            self._parseFile(profile_path, fd)
+
+    def _parseFile(self, _label, fd):
+
+        class ParseContext:
+            active_entries = []
+            label = _label
+
+        context = ParseContext()
+
+        for nr, line in enumerate(fd.readlines(), 1):
+            line = line.strip()
+
+            if not line or line.startswith('#'):
+                continue
+
+            context.line_nr = nr
+
+            self._parseLine(context, line)
+
+    def _parseLine(self, context, line):
+
+        if line.startswith('/') or line.startswith('%'):
+            context.active_entries = []
+
+            entry = PermissionsEntry(context.label, context.line_nr)
+            path, ownership, mode = line.split()
+            # the format supports both "user.group" and
+            # "user:group"
+            entry.owner, entry.group = ownership.replace('.', ':').split(':')
+            entry.mode = int(mode, 8)
+            expanded = self.m_var_handler.expandPaths(path)
+
+            for p in expanded:
+                entry.path = p
+                entry.related_paths = list(filter(lambda e: e != path, expanded))
+                key = entry.path.rstrip(os.path.sep)
+                if not key:
+                    # this is the root node, keep the slash
+                    key = '/'
+                entry_copy = copy.deepcopy(entry)
+                self.m_entries[key] = entry_copy
+                context.active_entries.append(entry_copy)
+        elif line.startswith('+'):
+            # capability line
+            _type, rest = line.split()
+            _type = _type.lstrip('+')
+
+            if _type != "capabilities":
+                raise Exception("Unexpected +[line] encountered in {}:{}".format(context.label, context.line_nr))
+
+            caps = rest.split(',')
```

Comment:
do we want some sanity checks here for the capabilities we allow? Prevent typos etc

Reply:
Currently the rpmlint checks don't perform any special sanity checks on the whitelists. I'd leave that to the whitelist checkers.